### PR TITLE
overlord,systemd: do not restart mount units on snapd start

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Do not restart the mount units when starting snapd even if the file
itself has been modified. This is to prevent systemd stopping services
when the mount unit gets restarted, as there are required dependencies
in the mount path for services defined in snaps.

### CODE REVIEW

**Review by valentindavid — APPROVED** [review]
Looks good. I have left some comments more for discussion. I think the PR is fine to merge if the tests pass.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, I've answered the comments and squashed the commits.

File: systemd/systemd_test.go [hunk 1]
```diff
@@ -1160,7 +1160,7 @@ func (s *SystemdTestSuite) TestAddMountUnit(c *C) {
 	mockSnapPath := filepath.Join(c.MkDir(), "/var/lib/snappy/snaps/foo_1.0.snap")
 	makeMockFile(c, mockSnapPath)
 
-	mountUnitName, err := NewUnderRoot(rootDir, SystemMode, nil).EnsureMountUnitFile("Mount unit for foo, revision 42", mockSnapPath, "/snap/snapname/123", "squashfs")
+	mountUnitName, err := NewUnderRoot(rootDir, SystemMode, nil).EnsureMountUnitFile("Mount unit for foo, revision 42", mockSnapPath, "/snap/snapname/123", "squashfs", &systemd.EnsureMountUnitFlags{})
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply systemd/systemd_test.go:1163]: Added now (see `testEnsureMountUnitChanged`)

File: systemd/emulation.go [hunk 2]
```diff
@@ -124,20 +124,21 @@ func (s *emulation) LogReader(services []string, n int, follow, namespaces bool)
 	return nil, fmt.Errorf("LogReader")
 }
 
-func (s *emulation) EnsureMountUnitFile(description, what, where, fstype string) (string, error) {
+func (s *emulation) EnsureMountUnitFile(description, what, where, fstype string, flags *EnsureMountUnitFlags) (string, error) {
```
Comment (zyga) [inline systemd/emulation.go:127]: Passing EnsureMountUnitFlags by value is probably better here. It's less worrying about escape analysis to optimize this away, less chance of a nil pointer.

File: systemd/systemd.go [hunk 3]
```diff
@@ -345,6 +348,15 @@ const (
 	mountCreated
 )
 
+// EnsureMountUnitFlags contains flags that modify behavior of EnsureMountUnitFile
+// TODO should we call directly EnsureMountUnitFileWithOptions and
+// remove this type instead?
+type EnsureMountUnitFlags struct {
+	// PreventRestartIfModified is set if we do not want to restart the
+	// mount unit if modified
```
Comment (zyga) [inline systemd/systemd.go:356]:
```suggestion
	// mount unit even thought it has been modified.```

**Review by zyga — APPROVED** [review]
LGTM with one comment about passing flags by value and one comment about wording of the documentation.

File: systemd/systemd.go [hunk 4]
```diff
@@ -1543,9 +1554,13 @@ func (s *systemd) EnsureMountUnitFileWithOptions(unitOptions *MountUnitOptions)
 		if err := s.EnableNoReload(units); err != nil {
 			return "", err
 		}
-		// Start/restart the created or modified unit now
-		if err := s.RestartNoWaitForStop(units); err != nil {
-			return "", err
+
+		// If just modified, some times it is not convenient to restart
+		if modified != mountUpdated || !unitOptions.IgnoreModified {
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply systemd/systemd.go]: `daemon-reload` does not do the remount. Maybe it is not well documented in systemd, but the mount unit has a `reload`. See https://github.com/systemd/systemd/blob/7b44a24c1bc764b5c0555fed08911665e1e7d446/src/core/mount.c#L2437

That calls in the end `mount_enter_remounting` https://github.com/systemd/systemd/blob/7b44a24c1bc764b5c0555fed08911665e1e7d446/src/core/mount.c#L1228

Which will use `-o remount`.

That said, `-o remount` does not support all changes. For example, you cannot change the device.

File: systemd/emulation.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply systemd/emulation.go]: Changed now

**Pull request merged**